lw/io/serializer/serializer.h

```diff
@@ -0,0 +1,317 @@
+#pragma once
+
+#include <concepts>
+#include <memory>
+#include <string_view>
+
+#include "lw/co/future.h"
+#include "lw/io/serializer/concepts.h"
+#include "lw/io/serializer/formatter.h"
+
+namespace lw::io {
+
+/**
+ * Tag to be used by specializations of `::lw::io::Serialize` to trigger object
+ * formatting in the serializer.
+ *
+ * For example, in JSON formatting, this tag results in `{` being inserted
+ * before calling the `Serialize::serialize` method, and `}` being inserted
+ * afterwards. When in object-formatting, the JSON formatter will also know to
+ * expect 2 values and insert a `:` between them along with a `,` before the 2nd
+ * through nth pairs.
+ *
+ * `Serialize` specializations with `typedef ObjectTag serialization_category`
+ * match the `ObjectSerializable` concept.
+ *
+ * @example{.cpp}
+ *  template<>
+ *  struct Serialize<MyObjectClass> {
+ *    typedef ObjectTag serialization_category; // <-- Object Tagged
+ *
+ *    void serialize(Serializer& serializer, const MyObjectClass&);
+ *    MyObjectClass deserialize(SerializedValue&);
+ *  };
+ */
+struct ObjectTag {};
+
+/**
+ * Tag to be used by specializations of `::lw::io::Serialize` to trigger list
+ * formatting in the serializer.
+ *
+ * For example, in JSON formatting, this tag results in `[` being inserted
+ * before calling the `Serialize::serialize` method, and `]` being inserted
+ * afterwards. When in list-formatting, the JSON formatter will also know to
+ * put a `,` before the 2nd through nth values.
+ *
+ * `Serialize` specializations with `typedef ListTag serialization_category`
+ * match the `ListSerializable` concept.
+ *
+ * @example{.cpp}
+ *  template<>
+ *  struct Serialize<MyListClass> {
+ *    typedef ListTag serialization_category; // <-- List Tagged
+ *
+ *    void serialize(Serializer& serializer, const MyListClass&);
+ *    MyListClass deserialize(SerializedValue&);
+ *  };
+ */
+struct ListTag {};
```

Comment:
Is this supposed to be a bracket pair?

Reply:
Yes. It is defining an empty struct. `ListTag` is struct (or class) with no data members or methods. This is a C++ idiom used to make checks at compile-time. The code can reference this struct to change what the compiler does without having any overhead on the final compiled binary.